include/rtio.h

```diff
@@ -0,0 +1,479 @@
+	 */
+	RTIO_TRIGGER_SOURCE_COUNT
+};
+
+/**
+ * @brief Trigger source configuration
+ *
+ * The source of the trigger may be an internal event along with the
+ * associated GPIO device and pin, a counter timeout, or a function call
+ */
+struct rtio_trigger_config {
+	enum rtio_trigger_source source;
+	union {
+		struct rtio_trigger_gpio {
+			struct device *gpio;
+			u32_t pin;
+		} gpio_config;
+		struct rtio_trigger_counter {
+			struct device *counter;
+			u32_t timeout;
+		} counter_config;
+	} options;
+};
+
+/**
+ * @brief Trigger event used in rtio_pipe.
+ */
+struct rtio_trigger_event {
+	u32_t trigger_source;
+	u32_t timestamp;
+};
+
+/**
+ * @brief Check if a block has met an output policy expectations
+ *
+ * @return true if the policy has been met, false otherwise
+ */
+__syscall bool rtio_output_policy_check(struct rtio_output_config *cfg,
+					struct rtio_block *block);
+
+static inline bool
+z_impl_rtio_output_policy_check(struct rtio_output_config *cfg,
+				struct rtio_block *block)
+{
+	if (k_cycle_get_32() - block->begin_tstamp > cfg->time) {
+		return true;
+	}
+	if (block->len > cfg->byte_size) {
+		return true;
+	}
+	return false;
+}
+
+/**
+ * @brief A rtio configuration
+ */
+struct rtio_config {
+	/* output configuration if applicable */
+	struct rtio_output_config *output_config;
+
+	/* trigger configuration if applicable */
+	struct rtio_trigger_config *trigger_config;
+
+	/* driver specific configuration */
+	void *driver_config;
+};
+
+/**
+ * @private
+ * @brief Function definition for configuring a RTIO device
+ */
+typedef int (*rtio_configure_t)(struct device *dev,
+				  struct rtio_config *config,
+				  u32_t *version);
+
+/**
+ * @private
+ * @brief Function definition for triggering a device read or write
+ */
+typedef int (*rtio_trigger_t)(struct device *dev);
+
+/**
+ * @brief Real-Time IO API
+ */
+struct rtio_api {
+	/* Configuration function pointer *must* be implemented */
+	rtio_configure_t configure;
+
+	/* Trigger function pointer *must* be implemented and IRQ safe */
+	rtio_trigger_t trigger;
+};
+
+/**
+ * @brief Configure the device
+ *
+ * This will reconfigure the device given a device specific configuration
+ * structure and possibly updates the block layout. This call will wait for
+ * any pending IO operations to be completed and prevent further IO operations
+ * until complete.
+ *
+ * Layout is assigned to the new buffer byte layout for the rtio device.
+ * Its expected the caller understands the new byte layout.
+ *
+ * @param dev RTIO device to configure
+ * @param config Configuration settings
+ * @param[out] layout New block layout id after reconfiguring
+ */
+__syscall int rtio_configure(struct device *dev,
+			     struct rtio_config *config,
+			     u32_t *layout);
+
+static inline int z_impl_rtio_configure(struct device *dev,
+					struct rtio_config *config,
+					u32_t *layout)
+{
+	const struct rtio_api *api = dev->driver_api;
+
+	if (!api->configure) {
```

Comment:
Totally minor, but it's documented that these must be implemented, so you could save a few bytes of code here by eliminating the check.